armi/reactor/components/component.py

```diff
@@ -582,30 +582,42 @@ def getNuclides(self):
 
         This includes anything that has been specified in here, including trace nuclides.
         """
-        return list(self.p.numberDensities.keys())
+        nucs = set(self.p.numberDensities.keys())
+        if self._children:
+            childNuclides = set(composites.Composite.getNuclides(self))
+            nucs |= childNuclides
+        return nucs
 
-    def getNumberDensity(self, nucName):
+    def getNuclideNumberDensities(self, nucNames):
         """
-        Get the number density of nucName, return zero if it does not exist here.
+        Return a list of number densities for the nuc names requested.
 
-        Parameters
-        ----------
-        nucName : str
-            Nuclide name
+        If this Component has children, then homogenize their number densities in.
 
-        Returns
-        -------
-        number density : float
-            number density in atoms/bn-cm.
+        Component ndens is unique in that it combines its own actual composition
+        with that of its potential children
         """
-        return self.p.numberDensities.get(nucName, 0.0)
-
-    def getNuclideNumberDensities(self, nucNames):
-        """Return a list of number densities for the nuc names requested."""
-        return [self.p.numberDensities.get(nucName, 0.0) for nucName in nucNames]
 
-    def _getNdensHelper(self):
-        return dict(self.p.numberDensities)
+        if self._children:
+            childDens = dict(
+                zip(
+                    nucNames,
+                    composites.Composite.getNuclideNumberDensities(self, nucNames),
+                )
+            )
+            # get volume of children by using the composite method which loops over all children
+            childVol = composites.Composite.getVolume(self)
+            # get self volume by calling the shape-specific volume method of the background shape
+            totalVol = self.getVolume()
+            childFrac = childVol / totalVol
+            selfFrac = 1.0 - childFrac
+            return [
+                self.p.numberDensities.get(nucName, 0.0) * selfFrac
+                + childDens.get(nucName, 0.0) * childFrac
+                for nucName in nucNames
+            ]
+        else:
+            return [self.p.numberDensities.get(nucName, 0.0) for nucName in nucNames]
```

Comment:
How would a plugin be able to get the number densities of just the matrix component with this change?

On one hand this is consistent with how the design of the composite tree: `Composite.getNuclideNumberDensities` gives the densities for all children on a node. So this would match well with `Block.getNuclideNumberDensities` for some nodal homogenization methods. But modeling this `Component` exactly in a monte carlo code would be problematic w/o the number densities of just the `Component` and not its children

Reply:
Great point. You would have to access the matrix's `.p.numberDensities` parameter directly in order to differentiate between it and its children with this change as written. Willing to consider more elegant options.

Reply:
@ntouran is

> matrix's .p.numberDensities

still the recommended approach here? I've been testing this out over the last few days and had a hard time remembering this thread.

Would something like `Component.getNuclideNumberDensities(children=False)` be acceptable? I tend to be hesitant pulling data from these parameter containers as the feel (maybe incorrectly) more private than a public method